src/lib/util/artifact_download.js

```diff
@@ -29,78 +31,58 @@ module.exports = async function(queue, stream, taskId, artifactPath, destination
     taskId,
     artifactPath
   );
-  let attempts = 0;
+
+  // As we change the semantics of these parameters, we add
+  // some preventive checks in case we didn't handle all the cases
+  // in production.
+  if (delayFactor > 10) {
+    delayFactor = RETRY_CONFIG.delayFactor;
+  }
+  if (typeof randomizationFactor !== 'boolean') {
+    randomizationFactor = RETRY_CONFIG.randomizationFactor;
+  }
 
   stream.write(
     fmtLog(`Downloading artifact "${artifactPath}" from task ID: ${taskId}.`)
   );
-  while (attempts++ < maxAttempts) {
-    let hash = crypto.createHash('sha256');
-    let destinationStream = fs.createWriteStream(destination);
-    try {
-      let expectedSize = 0;
-      let receivedSize;
-      let req = request.get(artifactUrl);
-      req.on('response', (res) => {
-        expectedSize = parseInt(res.headers['content-length']);
-        receivedSize = 0;
-      });
-      req.on('data', (chunk) => {
-        receivedSize += chunk.length;
-        hash.update(chunk);
-      });
 
-      let intervalId = setInterval(() => {
-        if (receivedSize) {
-          stream.write(fmtLog(
-            `Download Progress: ${((receivedSize / expectedSize) * 100).toFixed(2)}%`
-          ));
-        }
-      }, 5000);
-
-      try {
-        await pipe(req, destinationStream);
-      } finally {
-        clearInterval(intervalId);
-      }
+  let bar = new ProgressBar({
+    schema: 'Download progress :percent [:bar]',
+    filled: '=',
+    blank: ' ',
+  });
 
-      if (req.response.statusCode !== 200) {
-        let error = new Error(req.response.statusMessage);
-        error.statusCode = req.response.statusCode;
-        throw error;
-      }
+  return await promiseRetry((retry, attempt) => {
+    return new Promise(async(accept, reject) => {
```

Comment:
I think the Promise constructor ignores the return value of its argument.  In this case, that's another Promise (because it's an async function).  If anything in the body throws an exception, it will be converted to a rejection of that second Promise, which will go unhandled.

I think this could be

```javascript
return await promiseRetry(async (retry, attempt) => {
  try {
    let hash = crypto.createHash('sha256');
    // ..
    return `sha256:${hash.digest('hex')}`; 
  } catch (err) {
    debug(`Error downloading "${artifactPath}" from task ID "${taskId}". ${e}`);
    // ...
    retry(err);
  }
}, {
   retries: maxAttempts,
   factor: delayFactor,
   randomize: randomizationFactor,
});
```